xdsl/dialects/csl_/csl_stencil.py

```diff
@@ -0,0 +1,271 @@
+from collections.abc import Sequence
+from itertools import pairwise
+from typing import Annotated, cast
+
+from xdsl.dialects import builtin, memref, stencil
+from xdsl.dialects.builtin import IntegerAttr, IntegerType, TensorType
+from xdsl.dialects.experimental import dmp
+from xdsl.ir import Attribute, Dialect, Operation, ParametrizedAttribute, SSAValue
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    Operand,
+    ParameterDef,
+    attr_def,
+    irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    opt_attr_def,
+    opt_prop_def,
+    prop_def,
+    result_def,
+)
+from xdsl.parser import AttrParser, Parser
+from xdsl.printer import Printer
+from xdsl.traits import HasAncestor, Pure
+from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.hints import isa
+
+
+@irdl_attr_definition
+class ExchangeDeclarationAttr(ParametrizedAttribute):
+    """
+    A simplified version of dmp.exchange, from which it should be lowered
+    """
+
+    name = "csl_stencil.exchange"
+
+    neighbor_param: ParameterDef[builtin.DenseArrayBase]
+
+    def __init__(
+        self,
+        neighbor: Sequence[int] | builtin.DenseArrayBase,
+    ):
+        data_type = builtin.i64
+        super().__init__(
+            [
+                (
+                    neighbor
+                    if isinstance(neighbor, builtin.DenseArrayBase)
+                    else builtin.DenseArrayBase.from_list(data_type, neighbor)
+                ),
+            ]
+        )
+
+    @classmethod
+    def from_dmp_exch_decl_attr(cls, src: dmp.ExchangeDeclarationAttr):
+        return cls(src.neighbor)
+
+    @property
+    def neighbor(self) -> tuple[int, ...]:
+        data = self.neighbor_param.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string(f"<to {list(self.neighbor)}>")
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
+        parser.parse_characters("<")
+        parser.parse_characters("to")
+        to = parser.parse_comma_separated_list(
+            parser.Delimiter.SQUARE, parser.parse_integer
+        )
+        parser.parse_characters(">")
+
+        return [builtin.DenseArrayBase.from_list(builtin.i64, to)]
+
+
+@irdl_op_definition
+class PrefetchOp(IRDLOperation):
+    """
+    An op to indicate buffer prefetches.
+    """
```

Comment:
I would clarify that this is an op that talks about communication. I.e. that this represents a symmetrical exchange with neighbouring nodes.

In that sense I would strongly recommend naming this op similar to `dmp.swap`, as it is conceptuially quite close.

Reply:
That's a really good point actually. It might be easy for the symmetric nature of the communication to get lost in the word prefetch. At the same time, I wonder if `dmp.swap` semantics allow for more of a targeted point-to-point communication (i.e., true swapping) than CS hardware can actually support. I'll change the comment as suggested, and am not fully sure about the name right now.